README.md


# ChkMocks 
chkMocks is a basic R package for comparing experimental ZymoBiomics samples with theoretical composition. Default composition is for Product name D6300.Not tested with custom mocks.

ZymoBIOMICS™ Microbial Community Standard. [Catalog No. D6300](https://files.zymoresearch.com/protocols/_d6300_zymobiomics_microbial_community_standard.pdf)  
Fasta files for full-length 16S rRNA gene sequences downloaded from [ZymoBiomics](https://s3.amazonaws.com/zymo-files/BioPool/ZymoBIOMICS.STD.refseq.v2.zip)  

**Install**  

```

library(devtools) # make sure you have installed devtools
install_github("microsud/chkMocks")

```


The support for Non-Zymo Mock communities is experimental.  

### Acknowledgments        

* Medische Microbiologie en Infectiepreventie, University Medical Center Groningen, The Netherlands   
* RIVM, The Netherlands   


**Contact**   

Sudarshan A. Shetty [sudarshanshetty9[@]gmail[dot].com]  
 
 
 





NEWS.md
# chkMocks 0.1.02
 
* Added an Zymo Gut trainingset and phyloseq. 

# chkMocks 0.1.0

* Minor fixes to vignettes. 
* Added an example doc of MIB custom mocks. 

# chkMocks 0.1.0

* Added a `plotZymoDefault` function to the package.
* Added a `assignTaxonomyCustomMock` function to the package.
* Added a `compare2theorectical` function to the package.
* Added a `checkZymoBiomics` function to the package.
* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(chkMocks)



tests/testthat/test-checkZymoBiomics.R
test_that("checkZymoBiomics", {

  library(chkMocks)

  zym_sp <- c("Bacillus.subtilis","Listeria.monocytogenes",
              "Staphylococcus.aureus", "Enterococcus.faecalis",
              "Lactobacillus.fermentum","Escherichia.coli",
              "Salmonella.enterica","Pseudomonas.aeruginosa")
  taxa_names(ZymoExamplePseq) <- refseq(ZymoExamplePseq)
  myps <- checkZymoBiomics (ZymoExamplePseq,
                            mock_db = NULL,
                            multithread= 2,
                            minBoot = 80)
  mppseq <- myps$pseq
  tx.names <- taxa_names(mppseq)
  expect_equal(tx.names, zym_sp)

  cor.tb <- myps$corrTable

  vals <- ceiling(cor.tb$ZymoTheoretical*100)
  exp.vals <- c(86,79,79,87,86,51,64,56,54,75,57,45,66,61,50,57,54,52,60,100)
  expect_equal(exp.vals, exp.vals)

  })


R/data.R
#' @title ZymoBiomics
#'
#' @description ZymoBiomics data from ZymoBIOMICS Microbial Community Standard.
#'
#' @details ZymoBiomics data from ZymoBIOMICS Microbial Community Standard.
#'          Has a table with  information on Species and their theoretical
#'          composition in percentage values for Genomic DNA, 16S Only,
#'          16S and 18S, Genome Copy and Cell number. Product name D6300
#'
#' @name ZymoBiomics
#'
#' @docType data
#'
#' @author Sudarshan Shetty \email{sudarshanshetty9@gmail.com}
#'
#' @usage data(ZymoBiomics)
#'
#' @return Loads the data set in R.
#'
#' @format Phyloseq object
#'
#' @keywords Data
#'
#' @references
#'
#' \itemize{
#' \item{}{ZymoBIOMICS™ Microbial Community Standard.
#'        \emph{Catalog No. D6300.
#'        \url{https://files.zymoresearch.com/protocols/_d6300_zymobiomics_microbial_community_standard.pdf}
#' }
#' }
#' }
NULL

#' @title ZymoBiomicsFasta
#'
#' @description 16S and 18S sequences from ZymoBIOMICS Microbial Community Standard.
#'
#' @details ZymoBiomicsFasta has all 16S and 18S sequence data as \code{DNAStringSet}.
#'          Has 50 rDNA sequences, all copies of 16S rRNA gene are made available.
#'
#' @name ZymoBiomicsFasta
#'
#' @docType data
#'
#' @author Sudarshan Shetty \email{sudarshanshetty9@gmail.com}
#'
#' @usage data(ZymoBiomicsFasta)
#'
#' @return Loads the data set in R.
#'
#' @format DNAStringSet object
#'
#' @keywords Data
#'
#' @references
#'
#' \itemize{
#' \item{}{ZymoBIOMICS™ Microbial Community Standard.
#'        \emph{Catalog No. D6300.
#'        \url{https://files.zymoresearch.com/protocols/_d6300_zymobiomics_microbial_community_standard.pdf}
#'        \url{https://s3.amazonaws.com/zymo-files/BioPool/ZymoBIOMICS.STD.refseq.v2.zip}
#' }
#' }
#' }
NULL

#' @title ZymoBiomicsPseq
#'
#' @description Theoretical 16S composition of ZymoBIOMICS Microbial Community Standard.
#'
#' @details ZymoBiomicsPseq contains the \emph{theoretical} 16S composition of
#'          ZymoBIOMICS Microbial Community Standard. One sample, 8 taxa and 7
#'          taxonomic ranks.
#'
#' @name ZymoBiomicsPseq
#'
#' @docType data
#'
#' @author Sudarshan Shetty \email{sudarshanshetty9@gmail.com}
#'
#' @usage data(ZymoBiomicsPseq)
#'
#' @return Loads the data set in R.
#'
#' @format Phyloseq object
#'
#' @keywords Data
#'
#' @references
#'
#' \itemize{
#' \item{}{ZymoBIOMICS™ Microbial Community Standard.
#'        \emph{Catalog No. D6300.
#'        \url{https://files.zymoresearch.com/protocols/_d6300_zymobiomics_microbial_community_standard.pdf}
#' }
#' }
#' }
NULL

#' @title ZymoExamplePseq
#'
#' @description Contains ZymoBIOMICS Microbial Community Standard samples ran
#'              in different sequencing libraries as postive controls.
#'
#' @details ZymoExamplePseq contains the \emph{experimental} 16S composition of
#'          ZymoBIOMICS Microbial Community Standard. 9 samples, 946 ASVs and 7
#'          taxonomic ranks. ZymoBIOMIC mock community standards consisting of
#'          eight bacterial species \emph{Pseudomonas.aeruginosa},
#'          \emph{Escherichia coli}, \emph{Salmonella.enterica},
#'          \emph{Lactobacillus.fermentum}, \emph{Enterococcus.faecalis},
#'          \emph{Staphylococcus.aureus}, \emph{Listeria.monocytogenes},
#'          \emph{Bacillus.subtilis} and two fungal species
#'          \emph{Saccharomyces.cerevisiae} and \emph{Cryptococcus.neoformans}.
#'          The dilutions were done in microbial free water (Qiagen) in
#'          eight rounds of a serial three-fold dilution
#'          prior to DNA extraction. The data are for the V4 region of the
#'          16S rRNA gene profiled with Illumina MiSeq.
#'
#' @name ZymoExamplePseq
#'
#' @docType data
#'
#' @author Sudarshan Shetty \email{sudarshanshetty9@gmail.com}
#'
#' @usage data(ZymoExamplePseq)
#'
#' @return Loads the data set in R.
#'
#' @format Phyloseq object
#'
#' @keywords Data
#'
#' @references
#' \itemize{
#' \item{}{Karstens L, Asquith M, Davin S, Fair D, Gregory WT, Wolfe AJ,
#'         Braun J, McWeeney S. (2019) Controlling for contaminants in
#'         low-biomass 16S rRNA gene sequencing experiments.
#'        \emph{mSystems}, 4:e00290-19.
#'        \url{https://doi.org/10.1128/mSystems.00290-19}
#' }
#' }
#' Shetty SA 2021
NULL

#' @title ZymoBiomicsGutPseq
#'
#' @description Theoretical 16S composition of ZymoBIOMICS Microbial Community Standard.
#'
#' @details ZymoBiomicsGutPseq contains the \emph{theoretical} 16S composition of
#'          ZymoBIOMICS Gut Microbial Community Standard. One sample, 19 known taxa and 7
#'          taxonomic ranks.
#'
#' @name ZymoBiomicsGutPseq
#'
#' @docType data
#'
#' @author Sudarshan Shetty \email{sudarshanshetty9@gmail.com}
#'
#' @usage data(ZymoBiomicsGutPseq)
#'
#' @return Loads the data set in R.
#'
#' @format Phyloseq object
#'
#' @keywords Data
#'
#' @references
#' \itemize{
#' \item{}{ZymoBIOMICS® Gut Microbiome Standard.
#'        \emph{Catalog No. D6331.
#'        \url{https://files.zymoresearch.com/protocols/_d6331_zymobiomics_gut_microbiome_standard.pdf}
#' }
#' }
#' }
#' Shetty SA 2021
NULL

#' @title ZymoTrainingSet
#'
#' @description Contains ZymoTrainingSet compatible with DECIPHER::IdTaxa.
#'
#' @details ZymoTrainingSet contains 16S sequences of
#'          ZymoBIOMICS Microbial Community Standard. ZymoBIOMIC mock community
#'          standards consisting of
#'          eight bacterial species \emph{Pseudomonas.aeruginosa},
#'          \emph{Escherichia coli}, \emph{Salmonella.enterica},
#'          \emph{Lactobacillus.fermentum}, \emph{Enterococcus.faecalis},
#'          \emph{Staphylococcus.aureus}, \emph{Listeria.monocytogenes},
#'          \emph{Bacillus.subtilis} and two fungal species
#'          \emph{Saccharomyces.cerevisiae} and \emph{Cryptococcus.neoformans}.
#'
#' @name ZymoTrainingSet
#'
#' @docType data
#'
#' @author Sudarshan Shetty \email{sudarshanshetty9@gmail.com}
#'
#' @usage data(ZymoTrainingSet)
#'
#' @return Loads the data set in R.
#'
#' @format TrainingSet object
#'
#' @keywords Data
#'
#' @references
#' \itemize{
#' \item{}{ZymoBIOMICS™ Microbial Community Standard.
#'        \emph{Catalog No. D6300.
#'        \url{https://files.zymoresearch.com/protocols/_d6300_zymobiomics_microbial_community_standard.pdf}
#' }
#' }
#' }
#' Shetty SA 2021
NULL

#' @title ZymoBiomicsGutTrainingSet
#'
#' @description Contains ZymoBiomicsGutTrainingSet compatible with DECIPHER::IdTaxa.
#'
#' @details ZymoBiomicsGutTrainingSet contains 16S sequences of
#'          ZymoBIOMICS Gut Microbial Community Standard Catalog No. D6331.
#'
#' @name ZymoBiomicsGutTrainingSet
#'
#' @docType data
#'
#' @author Sudarshan Shetty \email{sudarshanshetty9@gmail.com}
#'
#' @usage data(ZymoBiomicsGutTrainingSet)
#'
#' @return Loads the data set in R.
#'
#' @format TrainingSet object
#'
#' @keywords Data
#'
#' @references
#' \itemize{
#' \item{}{ZymoBIOMICS® Gut Microbiome Standard.
#'        \emph{Catalog No. D6331.
#'        \url{https://files.zymoresearch.com/protocols/_d6331_zymobiomics_gut_microbiome_standard.pdf}
#' }
#' }
#' }
#' Shetty SA 2021
NULL

#' @title ZymoBiomicsGut
#'
#' @description Contains ZymoBiomicsGut compatible with DECIPHER::IdTaxa.
#'
#' @details ZymoBiomicsGut contains table with information on strains and thier
#'          abundances ZymoBIOMICS Gut Microbial Community Standard Catalog
#'          No. D6331.
#'
#' @name ZymoBiomicsGut
#'
#' @docType data
#'
#' @author Sudarshan Shetty \email{sudarshanshetty9@gmail.com}
#'
#' @usage data(ZymoBiomicsGut)
#'
#' @return Loads the data set in R.
#'
#' @format Table
#'
#' @keywords Data
#'
#' @references
#' \itemize{
#' \item{}{ZymoBIOMICS® Gut Microbiome Standard.
#'        \emph{Catalog No. D6331.
#'        \url{https://files.zymoresearch.com/protocols/_d6331_zymobiomics_gut_microbiome_standard.pdf}
#' }
#' }
#' }
#' Shetty SA 2021
NULL


R/plotZymoDefault.R
#' Plot ZymoBiomics to Experimental comparison
#'
#' @name plotZymoDefault
#'
#' @details A wrapper to plot the output of \code{checkZymoBiomics}.
#'
#' @param x Output from \code{checkZymoBiomics}
#'
#' @examples
#' library(phyloseq)
#' output.dat <- checkZymoBiomics(ZymoExamplePseq,
#'                                mock_db = ZymoTrainingSet,
#'                                multithread= 2,
#'                                threshold = 80,
#'                                strand = "top",
#'                                verbose = FALSE)
#' plotZymoDefault(output.dat)
#'
#' @author Sudarshan Shetty \email{sudarshanshetty9@gmail.com}
#'
#' @return ggplot object
#'
#' @importFrom dplyr %>% desc
#' @import patchwork
#'
#' @export

plotZymoDefault <- function(x){

  ZymoTheoretical <- abundance <- sample.chkmks <- chk.names <- FeatureID <- NULL
  if(class(x)=="list"
     && class(x[[1]])=="phyloseq"
     && class(x[[2]])=="phyloseq"
     && class(x[[3]])[3]=="data.frame") {


    ldf <- .get_long_tib(x$ps_species)

    # Correlation data
    cor.dat <- x$corrTable %>%
      dplyr::arrange(dplyr::desc(ZymoTheoretical))

    # for sorting sample names
    cor.sams <- x$corrTable %>%
      dplyr::arrange(desc(ZymoTheoretical)) %>%
      dplyr::pull(sample.chkmks)

    # sort sample names
    cor.dat$sample.chkmks <- factor(cor.dat$sample.chkmks, levels=cor.sams)
    ldf$chk.names <- factor(ldf$chk.names, levels = cor.sams)

    zym.cols <- c(Bacillus.subtilis = "#332D2D",
                  Listeria.monocytogenes ="#ECB333",
                  Staphylococcus.aureus="#76C6F3",
                  Enterococcus.faecalis="#009E73",
                  Lactobacillus.fermentum="#F0E442",
                  Escherichia.coli="#0072B2",
                  Salmonella.enterica="#EC6E0B",
                  Pseudomonas.aeruginosa="#C1066E",
                  na.value = "#bababa",
                  Unknown = "#bababa"

    )


    p1 <- ggplot2::ggplot(ldf,
                          ggplot2::aes(abundance, chk.names)) +
      ggplot2::geom_bar(
        ggplot2::aes(fill=FeatureID), stat = "identity") +
      ggplot2::scale_fill_manual(values = zym.cols) +
      ggplot2::theme_bw() +
      ggplot2::theme(legend.title = ggplot2::element_blank(),
                     legend.text = ggplot2::element_text(face="italic")) +
      ggplot2::xlab("Abundance (%)")+
      ggplot2::ylab("Samples")

    p2 <- ggplot2::ggplot(cor.dat,
                          ggplot2::aes(ZymoTheoretical,sample.chkmks)) +
      ggplot2::geom_col(ggplot2::aes(fill=ZymoTheoretical)) +
      ggplot2::xlab("Spearman's Correlation \nwith Theoretical") +
      ggplot2::theme_bw() +
      ggplot2::scale_fill_viridis_c("Pearson's Correlation") +
      ggplot2::theme(axis.text.y = ggplot2::element_blank(),
                     axis.ticks.y = ggplot2::element_blank(),
                     axis.title.y = ggplot2::element_blank(),
                     legend.position = "none") +
      ggplot2::geom_vline(xintercept = 1, lty=2)

    p1 + p2 + plot_layout(guides = "collect", widths = c(2,1))

  } else {

    stop("Not a valid input")

  }


}

#' @importFrom dplyr left_join
.get_long_tib <- function(x){

  FeatureID <- name <- value <- NULL
  tx.tib <- phyloseq::tax_table(x) %>%
    as.matrix() %>%
    as.data.frame() %>%
    tibble::rownames_to_column("FeatureID")
  otu.tib <- phyloseq::otu_table(x) %>%
    as.data.frame() %>%
    tibble::rownames_to_column("FeatureID") %>%
    tidyr::pivot_longer(!FeatureID)%>%
    dplyr::left_join(tx.tib, by="FeatureID") %>%
    dplyr::rename(chk.names=name,
                  abundance=value)

  sam_tib <- as(phyloseq::sample_data(x),"data.frame") %>%
    #as.data.frame(stringsAsFactors=FALSE) %>%
    tibble::rownames_to_column("chk.names") %>%
    tibble::as_tibble()

  ldf <- otu.tib %>%
    dplyr::left_join(sam_tib,by = "chk.names")

  return(ldf)
}


R/compare2theorectical.R
#' Compare Custom Mocks to Experimental
#'
#' @name compare2theorectical
#'
#' @details Correlation between theoretical and experimental mocks.
#' @param x Phyloseq object with experimental and theoretical composition
#' @param theoretical_id name of the theoretical sample. Must be in sample_names(x)
#'
#' @examples
#' library(phyloseq)
#' ZymoExamplePseq
#' compare2theorectical(ZymoExamplePseq, theoretical_id = "somename")
#'
#' @author Sudarshan Shetty \email{sudarshanshetty9@gmail.com}
#'
#' @return Correlation table
#'
#' @importFrom dplyr %>% all_of
#' @importFrom tibble as_tibble
#' @importFrom phyloseq otu_table sample_names
#' @importFrom corrr correlate focus
#'
#' @export

compare2theorectical <- function(x, theoretical_id){

  otu.tb <- phyloseq::otu_table(x) %>%
    as.matrix() %>%
    as.data.frame() %>%
    tibble::as_tibble()

  if(!theoretical_id %in% sample_names(x) || is.null(theoretical_id)){
    message(paste0("Check that theoretical = ", theoretical_id, " exists in sample_names(x)"))
    message("returning all pairwise comparisons")
    tex_cor <- suppressMessages(corrr::correlate(otu.tb,
                                                 method = "spearman",
                                                 use = 'everything'))
    return(tex_cor)

  } else {


    #column_id <- sym(column_id)


    tex_cor <- suppressMessages(corrr::correlate(otu.tb,
                                                 method = "spearman",
                                                 use = 'all.obs')) %>%
      corrr::focus(dplyr::all_of(theoretical_id))
    return(tex_cor)
  }


}


R/assignTaxonomyCustomMock.R
#' Assign Custom Mocks Taxonomy
#'
#' @name assignTaxonomyCustomMock
#'
#' @details For consistency in taxonomic assignments, use only a small database
#'          that has 16S rRNA gene sequences from mock community genomes.
#'          This is done to avoid any clashes/changes that happen with public databases
#'          and to get a direct comparison.
#'
#' @param x Phyloseq objects where taxa_names are ASV sequences
#'
#' @param mock_db ZymoTrainingSet
#'
#' @param processors Passed on to DECIPHER::IdTaxa Default is NULL
#'
#' @param threshold Passed on to DECIPHER::IdTaxa. Default is 60
#'
#' @param verbose Default is FALSE
#'
#' @param strand DECIPHER::IdTaxa Default is top
#'
#' @param ... Arguments to pass on to DECIPHER::IdTaxa
#'
#' @examples
#' #output.dat <- assignTaxonomyCustomMock(ps.zym,
#'  #                                      mock_db = NULL,
#'   #                                     processors = 2,
#'    #                                    threshold = 80)
#' @author Sudarshan Shetty \email{sudarshanshetty9@gmail.com}
#'
#' @return Phyloseq with new taxonomy
#'
#' @importFrom DECIPHER IdTaxa
#' @importFrom Biostrings DNAStringSet
#' @importFrom phyloseq taxa_names tax_table
#' @importFrom tidyr separate
#' @importFrom dplyr select
#'
#' @export
#' @importFrom dada2 assignTaxonomy
assignTaxonomyCustomMock <- function(x,
                                     mock_db = NULL,
                                     processors = NULL,
                                     threshold = 60,
                                     strand = "top",
                                     verbose = FALSE, ...){

  if(is.null(mock_db)){

    stop("Please provide reference trainingset")

  }

  Classification  <- root <- NULL
  dna <- Biostrings::DNAStringSet(taxa_names(x)) # Create a DNAStringSet from the ASVs

  ids <- DECIPHER::IdTaxa(dna, mock_db, strand=strand,
                          processors=processors,
                          threshold = threshold,
                          verbose = verbose)

  ranks <- c("Root","Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species") # ranks of interest
  assignment <- sapply(ids,
                       function(x)
                         paste(x$taxon,
                               collapse=";"))

  #class(assignment)

  ranks <- c("root","domain", "phylum", "class", "order", "family", "genus", "species") # ranks of interest
  custom_taxa <- data.frame(row.names = taxa_names(x),
                            Classification=assignment) %>%
    tidyr::separate(Classification, ranks, ";") %>%
    dplyr::select(!root) %>%
    as.matrix() %>%
    phyloseq::tax_table()

  # add new taxonomy
  phyloseq::tax_table(x) <- custom_taxa

  return(x)

}





R/checkZymoBiomics.R
#' Compare ZymoBiomics to Experimental
#'
#' @name checkZymoBiomics
#'
#' @details Compare ZymoBiomics theoretical to Experimental composition.
#'
#' @param x Phyloseq objects where taxa_names are ASV sequences
#'
#' @param mock_db ZymoTrainingSet
#'
#' @param multithread Passed on to DECIPHER::IdTaxa Default is NULL
#'
#' @param threshold Passed on to DECIPHER::IdTaxa. Default is 60
#'
#' @param verbose Default is FALSE
#'
#' @param strand DECIPHER::IdTaxa Default is top
#'
#' @param ... Arguments to pass on to DECIPHER::IdTaxa
#'
#' @examples
#' library(phyloseq)
#'
#' output.dat <- checkZymoBiomics(ZymoExamplePseq,
#'                                mock_db = ZymoTrainingSet,
#'                                multithread= 2,
#'                                threshold = 80,
#'                                strand = "top",
#'                                verbose = FALSE)
#'
#' @author Sudarshan Shetty \email{sudarshanshetty9@gmail.com}
#'
#' @return List with pseq, and cor table
#'
#' @importFrom dplyr %>%
#' @importFrom tibble as_tibble
#'
#' @export
#'

checkZymoBiomics <- function(x,
                             mock_db = NULL,
                             multithread= 2,
                             threshold = 60,
                             verbose = FALSE,
                             strand = "top",
                             ...) {

  if(is.null(mock_db)){
    stop("Specifiy `mock_db` ZymoTrainingSet")
  }
  sample.chkmks <- ZymoTheoretical <- row.sample <- NULL

  new_tx <- .assign_zymo_taxonomy(x,
                                  mock_db = mock_db,
                                  processors=multithread,
                                  threshold = threshold,
                                  strand = strand,
                                  verbose = verbose)

  nw_x <- .get_species_comp(new_tx)

  # merge with theoretical
  phyloseq::sample_data(ZymoBiomicsPseq)$ZymoType <- "Theoretical"
  phyloseq::sample_data(nw_x)$ZymoType <- "UserSamples"

  zm.all.ps <- phyloseq::merge_phyloseq(nw_x, ZymoBiomicsPseq)
  # Check correlation
  tex_cor <- .check_zym_cor(zm.all.ps)
  tex_cor <- tex_cor %>%
    tibble::add_row(sample.chkmks= "ZymoTheoretical", ZymoTheoretical = 1) %>%
    dplyr::mutate(row.sample = stats::reorder(sample.chkmks, ZymoTheoretical))

  return(list("ps_asv"=new_tx,
              "ps_species"=zm.all.ps,
              "corrTable" = tex_cor))

}

#' @importFrom DECIPHER IdTaxa
#' @importFrom Biostrings DNAStringSet
#' @importFrom phyloseq taxa_names tax_table
#' @importFrom tidyr separate
#' @importFrom dplyr select
.assign_zymo_taxonomy <- function(x,
                                  mock_db = mock_db,
                                  processors = processors,
                                  threshold = threshold,
                                  strand = strand,
                                  verbose = verbose, ...){

  message("Using internal reference database
            for ZymoBiomics")

  Classification  <- root <- NULL
  dna <- Biostrings::DNAStringSet(taxa_names(x)) # Create a DNAStringSet from the ASVs

  ids <- DECIPHER::IdTaxa(dna, mock_db, strand=strand,
                          processors=processors,
                          threshold = threshold,
                          verbose = verbose)

  ranks <- c("Root","Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species") # ranks of interest
  assignment <- sapply(ids,
                       function(x)
                         paste(x$taxon,
                               collapse=";"))

  #class(assignment)

  ranks <- c("root","domain", "phylum", "class", "order", "family", "genus", "species") # ranks of interest
  zym_taxa <- data.frame(row.names = taxa_names(x),
                         Classification=assignment) %>%
    tidyr::separate(Classification, ranks, ";") %>%
    dplyr::select(!root) %>%
    as.matrix() %>%
    phyloseq::tax_table()

  # add new toaxonomy
  phyloseq::tax_table(x) <- zym_taxa

  return(x)

}


#' @importFrom microbiome transform aggregate_taxa
#' @importFrom phyloseq otu_table otu_table<-
.get_species_comp <- function(x){

  x <- microbiome::aggregate_taxa(x, "species")
  x <- microbiome::transform(x, "compositional")
  phyloseq::otu_table(x) <- phyloseq::otu_table(x) *100
  return(x)

}

#' @importFrom phyloseq otu_table otu_table<-
#' @importFrom corrr correlate focus
#' @importFrom dplyr %>%
.check_zym_cor <- function(x){

  ZymoTheoretical <- NULL
  otu.tb <- phyloseq::otu_table(x) %>%
    as.matrix() %>%
    as.data.frame() %>%
    tibble::as_tibble()

  tex_cor <- suppressMessages(corrr::correlate(otu.tb,
                                               method = "spearman",
                                               use = 'all.obs')) %>%
    corrr::focus(ZymoTheoretical) %>%
    dplyr::rename(sample.chkmks="term")

  return(tex_cor)

}


